### PR TITLE
fix render_test_result step dependency

### PR DESCRIPTION
**Pull request by walterddr**
render_test_results should depend on successful steps of:
- calculate-docker-image
- build

If either of the step above failed it should not re run. 

The goal here is:
1. Reduce CI cost - even the step only runs for around 1min with just the checkout step and artifact download before failing, we still spend unnecessary time/cost
2. render_test_results failure contributes to "failing" count to the PR page signal hub - which is misleading when debugging failures

### CODE REVIEW

**Review by samestep — COMMENTED** [review]
I don't see what this accomplishes, since the `render_test_results` step already has `if: always()`?

**Comment by samestep** [comment]
I think you could make this do what you want by (in addition to adding those two lines to `needs`) also changing the `if` condition so that it only runs if those steps succeed

File: .github/templates/linux_ci_workflow.yml.in [hunk 1]
```diff
@@ -243,6 +243,8 @@ jobs:
   render_test_results:
     if: always()
     needs:
+      - calculate-docker-image
+      - build
```
Comment (seemethere) [inline .github/templates/linux_ci_workflow.yml.in:247]: Isn't there an intrinsic dependency within this? Like if you depend on `test` which depends on `build` and `calculate-docker-image` wouldn't it just auto depend on the dependencies of its dependencies?

**Review by seemethere — CHANGES_REQUESTED** [review]

**Comment by seemethere** [comment]
> I think you could make this do what you want by (in addition to adding those two lines to `needs`) also changing the `if` condition so that it only runs if those steps succeed

It should always run because it should also run if `test` fails so that it renders the failed result as well

File: .github/templates/linux_ci_workflow.yml.in (hunk 1 above)
Comment (samestep) [reply .github/templates/linux_ci_workflow.yml.in:247]: yeah exactly; see my comments above

**Comment by samestep** [comment]
> It should always run because it should also run if `test` fails so that it renders the failed result as well

right, but it doesn't need to run if the `calculate-docker-image` or `build` steps fail

File: .github/templates/linux_ci_workflow.yml.in (hunk 1 above)
Comment (walterddr) [reply .github/templates/linux_ci_workflow.yml.in]: fixed.

**Comment by samestep** [comment]
could you give a test plan for this? perhaps open some [do not merge] PRs on top of this to test

**Comment by walterddr** [comment]
> could you give a test plan for this? perhaps open some [do not merge] PRs on top of this to test

thanks for the comment. I dont think the PR at the current stage is ready for review based on your feedbacks. converting it into a draft first before adding a test plan

**Comment by seemethere** [comment]
> > It should always run because it should also run if `test` fails so that it renders the failed result as well
> 
> right, but it doesn't need to run if the `calculate-docker-image` or `build` steps fail

But there isn't really a harm in running it if those fail though?

**Comment by samestep** [comment]
> But there isn't really a harm in running it if those fail though?

yeah this is my question too

**Comment by walterddr** [comment]
I think the benefit I am looking for here is:
1.  even the step only runs for around 1min with just the checkout step and artifact download before failing, we still spend unnecessary time/cost
2. it contributes to "failing" count to the PR page signal hub which is misleading.

**Comment by walterddr** [comment]
approach didn't work and dont see an easy way to do it. going to close this and create an issue.

File: .github/templates/linux_ci_workflow.yml.in [hunk 2]
```diff
@@ -241,7 +241,7 @@ jobs:
           # Prune all of the docker images
           docker system prune -af
   render_test_results:
-    if: always()
+    if: always() && jobs.calculate-docker-image.outcome == 'success' && jobs.build.outcome == 'success'
```
Comment (malfet) [inline .github/templates/linux_ci_workflow.yml.in:244]: Hmm, we explicitly want to run this job when `job.build.outcome == 'failure'`, but not when it was skipped

Comment (walterddr) [reply .github/templates/linux_ci_workflow.yml.in:244]: i think what we want is react to the **test** job: `job.test.outcome != 'skipped'` 
if build job failed. the test outcome will guarantee to be empty right?

Comment (walterddr) [reply .github/templates/linux_ci_workflow.yml.in:244]: alternative that @samestep suggested is to move the download artifact as the very first step of the job, and mark job as successful if no test report artifact exists. (i dunno if there's a way to mark a job as skipped, related discussion: https://gitlab.com/gitlab-org/gitlab/-/issues/15266)

Comment (samestep) [reply .github/templates/linux_ci_workflow.yml.in:244]: the idea would be similar to this #56071

**Pull request closed without merging**